### PR TITLE
Issue #448 make state thread safe in consumer_partition and connection

### PR DESCRIPTION
**Pull request by zzzming**
### Motivation
Address Issue: #448 and connection.go also missing a few sync against state

### Modifications

Use go sync/atomic to synchronize state and add to missing ones in connection.go

I did not use uber's atomic as suggested by @flowchartsman. The sync/atomic seems adequate and avoid another dependency. If uber's atomic is required, we should update other sync/atomic to be consistent. @merlimat @wolfstudy what do you think? thanks

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: no
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no

### Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented?
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by flowchartsman** [comment]
I'm not particularly attached to uber's atomic, I used it for expediency's sake, but direct calls to atomic always make me nervous if I'm doing them from multiple places, so  guarding it in a (probably inlined) method is usually the way I like to go. YMMV. But, if it works and it's consistent with other code, who am I to judge? ¯\_ʕ◔ϖ◔ʔ_/¯ Aside from that, LGTM

**Comment by zzzming** [comment]
> I'm not particularly attached to uber's atomic, I used it for expediency's sake, but direct calls to atomic always make me nervous if I'm doing them from multiple places, so guarding it in a (probably inlined) method is usually the way I like to go. YMMV. But, if it works and it's consistent with other code, who am I to judge? ¯_ʕ◔ϖ◔ʔ_/¯ Aside from that, LGTM

@flowchartsman I share the same sentiment that I would tend to forget about add sync.atomic too. Let's wait for Matteo to weight in. It's a minor change either way. cheers

**Comment by merlimat** [comment]
The `uber/atomic` looks indeed more readable and safer than the `sync/atomic`. It would be good to use that consistently

**Comment by zzzming** [comment]
> The `uber/atomic` looks indeed more readable and safer than the `sync/atomic`. It would be good to use that consistently

updated all states with uber/atomic

File: pulsar/producer_partition.go [hunk 1]
```diff
@@ -703,8 +707,22 @@ func (p *partitionProducer) Flush() error {
 	return cp.err
 }
 
+func (p *partitionProducer) getProduerState() producerState {
```
Comment (merlimat) [inline pulsar/producer_partition.go:710]:
```suggestion
func (p *partitionProducer) getProducerState() producerState {
```

File: pulsar/producer_partition.go [hunk 2]
```diff
@@ -263,7 +267,7 @@ func (p *partitionProducer) reconnectToBroker() {
 	}
 
 	for maxRetry != 0 {
-		if atomic.LoadInt32(&p.state) != producerReady {
+		if p.getProduerState() != producerReady {
```
Comment (merlimat) [inline pulsar/producer_partition.go:270]:
```suggestion
		if p.getProducerState() != producerReady {
```

File: pulsar/producer_partition.go [hunk 3]
```diff
@@ -443,7 +447,7 @@ func (p *partitionProducer) internalFlushCurrentBatch() {
 
 func (p *partitionProducer) failTimeoutMessages() {
 	// since Closing/Closed connection couldn't be reopen, load and compare is safe
-	state := atomic.LoadInt32(&p.state)
+	state := p.getProduerState()
```
Comment (merlimat) [inline pulsar/producer_partition.go:450]:
```suggestion
	state := p.getProducerState()
```

File: pulsar/producer_partition.go [hunk 4]
```diff
@@ -703,8 +707,22 @@ func (p *partitionProducer) Flush() error {
...
+
+// set a new consumerState and return the last state
+// returns bool if the new state has been set or not
+func (p *partitionProducer) casProducerState(oldState, newState producerState) bool {
+	return p.state.CAS(int32(oldState), int32(newState))
+}
+
 func (p *partitionProducer) Close() {
-	if atomic.LoadInt32(&p.state) != producerReady {
+	if p.getProduerState() != producerReady {
```
Comment (merlimat) [inline pulsar/producer_partition.go:725]:
```suggestion
	if p.getProducerState() != producerReady {
```

**Review by merlimat — APPROVED** [review]

**Pull request merged**